File: lib/java/openlinktoken/src/main/java/org/openlinktoken/tokens/tokenizer/TokenDigestFactory.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.tokens.tokenizer;

import java.security.NoSuchAlgorithmException;

import org.openlinktoken.crypto.CryptoSuite;

/**
 * Creates token digest implementations from suite or algorithm identifiers.
 */
public final class TokenDigestFactory {

    /**
     * Prevents construction of this factory.
     */
    private TokenDigestFactory() {
    }

    /**
     * Creates the digest implementation declared by a crypto suite.
     *
     * @param cryptoSuite the suite selecting the digest
     * @return the selected digest implementation
     */
    public static TokenDigest forSuite(CryptoSuite cryptoSuite) {
        if (cryptoSuite == null) {
            throw new IllegalArgumentException("A valid CryptoSuite is required to select a token digest.");
        }
        try {
            return forAlgorithm(cryptoSuite.getTokenDigestAlgorithm());
        } catch (NoSuchAlgorithmException error) {
            throw new IllegalArgumentException(
                    "Unsupported token digest algorithm '" + cryptoSuite.getTokenDigestAlgorithm() + "'.",
                    error);
        }
    }

    /**
     * Creates a digest implementation from its registered algorithm identifier.
     *
     * @param algorithm the suite-level digest identifier
     * @return the selected digest implementation
     * @throws NoSuchAlgorithmException if the identifier is unsupported
     */
    public static TokenDigest forAlgorithm(String algorithm) throws NoSuchAlgorithmException {
        return switch (algorithm) {
            case "SHA-256" -> new SHA256TokenDigest();
            case "SHA3-256" -> new SHA3TokenDigest();
            case "SHAKE256-256" -> new SHAKE256TokenDigest();
            default -> throw new NoSuchAlgorithmException("Unsupported token digest algorithm '" + algorithm + "'.");
        };
    }
}

File: lib/python/openlinktoken/src/main/openlinktoken/tokens/tokenizer/token_digest_factory.py
# SPDX-License-Identifier: MIT

from openlinktoken.crypto_suite import CryptoSuite
from openlinktoken.tokens.tokenizer.sha3_token_digest import Sha3TokenDigest
from openlinktoken.tokens.tokenizer.sha256_token_digest import Sha256TokenDigest
from openlinktoken.tokens.tokenizer.shake256_token_digest import Shake256TokenDigest
from openlinktoken.tokens.tokenizer.token_digest import TokenDigest


class TokenDigestFactory:
    """Create token digest implementations from validated suite metadata."""

    _IMPLEMENTATIONS = {
        "SHA-256": Sha256TokenDigest,
        "SHA3-256": Sha3TokenDigest,
        "SHAKE256-256": Shake256TokenDigest,
    }

    @classmethod
    def for_suite(cls, crypto_suite: CryptoSuite) -> TokenDigest:
        """Create the digest implementation declared by a crypto suite."""
        if not isinstance(crypto_suite, CryptoSuite):
            raise ValueError("A valid CryptoSuite is required to select a token digest.")
        return cls.for_algorithm(crypto_suite.token_digest_algorithm)

    @classmethod
    def for_algorithm(cls, algorithm: str) -> TokenDigest:
        """Create a digest implementation from its registered algorithm name."""
        try:
            return cls._IMPLEMENTATIONS[algorithm]()
        except KeyError as error:
            raise ValueError(f"Unsupported token digest algorithm '{algorithm}'.") from error
